types/tx.go

```diff
@@ -102,6 +92,221 @@ func ToTxs(txs [][]byte) Txs {
+		case abci.TxRecord_REMOVED:
+			txrSet.removed = append(txrSet.removed, txrSet.all[i])
+		}
+	}
+	return txrSet
+}
+
+// GetAddedTxs returns the transactions marked for inclusion in a block. This
+// list maintains the order that the transactions were included in the list of
+// TxRecords that were used to construct the TxRecordSet.
+func (t TxRecordSet) GetIncludedTxs() []Tx {
+	return t.included
+}
+
+// GetAddedTxs returns the transactions added by the application.
+func (t TxRecordSet) GetAddedTxs() []Tx {
+	return t.added
+}
+
+// GetRemovedTxs returns the transactions marked for removal by the application.
+func (t TxRecordSet) GetRemovedTxs() []Tx {
+	return t.removed
+}
+
+// Validate checks that the record set was correctly constructed from the original
+// list of transactions.
+func (t TxRecordSet) Validate(maxSizeBytes int64, otxs Txs) error {
+	if len(t.unknown) > 0 {
+		return fmt.Errorf("transaction incorrectly marked as unknown, transaction hash: %x", t.unknown[0].Hash())
+	}
+
+	// The following validation logic performs a set of sorts on the data in the TxRecordSet indexes.
+	// It sorts the original transaction list, otxs, once.
+	// It sorts the new transaction list twice: once when sorting 'all', the total list,
+	// and once by sorting the set of the added, removed, and unmodified transactions indexes,
+	// which, when combined, comprise the complete list of transactions.
+	//
+	// The original list is iterated once and the modified list and set of indexes are
+	// also iterated once each.
+	// Asymptotically, this yields a total runtime of O(N*log(N) + N + 2*M*log(M) + 2*M),
+	// in the input size of the original list and the input size of the new list respectively.
+
+	// make a copy of the all index so that the original order of the all index
+	// can be preserved.
+	// does not copy the underlying data.
+	allCopy := make([]Tx, len(t.all))
+	copy(allCopy, t.all)
+	sort.Sort(Txs(allCopy))
+
+	var size int64
+	for i := 0; i < len(allCopy); i++ {
+		size += int64(len(allCopy[i]))
+		if size > maxSizeBytes {
+			return fmt.Errorf("transaction data size %d exceeds maximum %d", size, maxSizeBytes)
+		}
+
+		// allCopy is sorted, so any duplicated data will be adjacent.
+		if i < len(allCopy)-1 && bytes.Equal(allCopy[i], allCopy[i+1]) {
+			return fmt.Errorf("TxRecords contains duplicate transaction, transaction hash: %x", allCopy[i].Hash())
+		}
+	}
+
+	// create copies of each of the action-specific indexes so that order of the original
+	// indexes can be preserved.
+	addedCopy := make([]Tx, len(t.added))
+	copy(addedCopy, t.added)
+	removedCopy := make([]Tx, len(t.removed))
+	copy(removedCopy, t.removed)
+	unmodifiedCopy := make([]Tx, len(t.unmodified))
+	copy(unmodifiedCopy, t.unmodified)
+
+	sort.Sort(otxs)
+	sort.Sort(Txs(addedCopy))
+	sort.Sort(Txs(removedCopy))
+	sort.Sort(Txs(unmodifiedCopy))
+	unmodifiedIdx, addedIdx, removedIdx := 0, 0, 0
+	for i := 0; i < len(otxs); i++ {
```

Comment:
I feel as if this might be a little easier to read if instead of trying to fold them all together we did it as three loops: One for removed, one for unmodified. Instead of one big loop over otxs, each would visit otxs for just its own.

That _would_ mean potentially three passes, but each one would be self-contained, and at most one of the three would consume the entire slice. The legibility win might be worth it (and the traversals are cheap—we're not moving any data around).

Another option, since we already have them in order, would be to bsearch. We've already spent n lg n, so this is only constant overhead, and could be pulled into a helper:

```go
func findBytes(needle []byte, haystack [][]byte) (int, bool) {
   n := sort.Search(len(haystack), func(i int) bool {
      return bytes.Compare(haystack[i], needle) <= 0
   })
   if n == len(haystack) || !bytes.Equal(haystack[n], needle) {
      return 0, false
   }
   return n, true
}
```

(warning: untested but should be basically right)